### PR TITLE
feat: 🎸 methods and db changes to publish single map and batch-publish multiple maps (no interface yet)

### LINKED ISSUE #1771: Create batch-publish and single publish endpoints
Routes for publishing a single map or multiple maps at once should be created and extensively tested.

### PR DESCRIPTION
**Pull request by PeculiarE**
Fixes #1771

* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts 📁
* [ ] screenshots/GIFs are attached 📎 in case of UI updation
* [x] ask `@publiclab/mapknitter-reviewers` for help, in a comment below

### CODE REVIEW

**Comment by PeculiarE** [comment]
> @PeculiarE Seems like Rubocop has a few complaints about the comments. Not really important.

Yeah. I've fixed it though in the latest commit I just pushed.

File: app/controllers/spam_controller.rb [hunk 1]
```diff
@@ -43,4 +53,30 @@ def batch_spam_map
...
+    published_maps = 0
+    unbanned_authors = Set.new
+    params[:ids].split(',').uniq.each do |id|
+      map = Map.find(id)
+      if publish_check(map)
+        published_maps += 1
+        unbanned_authors << map.user.id if unban_check(map)
+      end
+    end
+    flash[:notice] = helpers.pluralize(published_maps, 'map') + ' published and ' + helpers.pluralize(unbanned_authors.size, 'author') + ' unbanned.'
```
[the remark this answers is not in the document]
Comment (PeculiarE) [reply app/controllers/spam_controller.rb]: Yeah, I totally get now. Made the change. Does this work fine?

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -53,7 +53,7 @@ def logged_in?
 
   def logged_in_as(roles, action)
     unless current_user && roles.any? { |role| current_user.role == role }
-      flash[:error] = "Only #{roles.collect { |role| role.pluralize }.join(' and ')} can #{action}."
+      flash[:error] = "Only #{roles.collect(&:pluralize).join(" and ")} can #{action}."
```
[the remark this answers is not in the document]
Comment (jywarren) [reply app/controllers/application_controller.rb:56]: ooh cool!

File: app/controllers/spam_controller.rb [hunk 3]
```diff
@@ -1,20 +1,28 @@
...
       map.spam unless map.status == Map::Status::BANNED
     end
   
     def ban_check(map)
-      #check and ban only unbanned non-anonymous authors
+      # check and ban only unbanned non-anonymous authors
       map.user.ban unless map.anonymous? || map.user.status == User::Status::BANNED
     end
+
+    def publish_check(map)
```
Comment (jywarren) [inline app/controllers/spam_controller.rb:13]: Do you think it's worth using a longer method name that reads more like human speech, like `check_and_publish`? Just an idea. The comments make it clear but i'm thinking when it's used in other files.

File: app/controllers/spam_controller.rb [hunk 4]
```diff
@@ -43,4 +53,30 @@ def batch_spam_map
     flash[:notice] = helpers.pluralize(spammed_maps, 'map') + ' spammed and ' + helpers.pluralize(banned_authors.size, 'author') + ' banned.'
     redirect_back(fallback_location: root_path)
   end
+
+  def publish_map
+    @map = Map.find(params[:id])
+    if publish_check(@map)
+      notice_text = 'Map published.'
+      notice_text.chop! << ' and author unbanned.' if unban_check(@map)
```
[the remark this answers is not in the document]
Comment (jywarren) [reply app/controllers/spam_controller.rb:59]: yes, very nice!!!

File: test/models/user_test.rb [hunk 5]
```diff
@@ -52,11 +52,19 @@ class UserTest < ActiveSupport::TestCase
...
+    assert_nil user.status_updated_at
+
     user.ban
     assert_equal User::Status::BANNED, user.status
+    assert_not_nil user.status_updated_at
+    old_time = user.status_updated_at
+
+    user.unban
+    assert_equal User::Status::NORMAL, user.status
+    assert user.status_updated_at.to_f >= old_time.to_f
```
Comment (jywarren) [inline test/models/user_test.rb:67]: Love all the tests, very thorough. Thank you!

**Review by jywarren — COMMENTED** [review]
Just a thought on method naming. If you prefer to just leave as-is, let me know! Otherwise are we ready to merge? Thank you!!!

File: test/models/user_test.rb (hunk 5 above)
Comment (PeculiarE) [reply test/models/user_test.rb:67]: Thank you 😄 !

File: app/controllers/spam_controller.rb (hunk 3 above)
Comment (PeculiarE) [reply app/controllers/spam_controller.rb:13]: Yeah, it's definitely worth it. I'm all for longer method-names for better readability. Will use your suggestions and re-push. Thank you ❤️

**Comment by PeculiarE** [comment]
> Just a thought on method naming. If you prefer to just leave as-is, let me know! Otherwise are we ready to merge? Thank you!!!

I've renamed the methods, and I believe we're ready to merge 😄...Thank you!

**Comment by jywarren** [comment]
Looks super, thank you!!!

**Pull request merged**